Add `release.yml` - set changelog entry categories

tomalec opened the pull request:
### Changes proposed in this Pull Request:

Add a simple release config, to use automatically (https://github.com/woocommerce/grow/pull/14) or manually set PR labels to categorize release notes' entries.


### Screenshots:

<!--- Optional --->


### Detailed test instructions:
<!-- Add detailed instructions for how to test that this PR fixes the issue and confirm that it doesn't break any other features :) -->
It could be hard to test this config, without merging it and merging a few other PRs after.
You can take a look at a test repo at https://github.com/woorelease-bugs/past-release/ which uses the same setup.

1. Open https://github.com/woorelease-bugs/past-release/releases/edit/1.0.1 
2. Click "Generate release notes"
3. Fork the repo
4. Create a new PR, attach `changelog: fix` label (or any other mentioned in the config)
5. Merge the PR
6. Generate new release notes as in 1-2.


### Additional details:

- This PR uses a new set of categories. The one that was agreed on below. That amis to unify but is different from current
   - set of changelog entry types (from PR template) `(Fix|Add|Tweak|Update) - `
   - branches name prefixes mentioned in wiki and P2 `(add/fix/update/remove)/(issue-number)-short-slug`
- Once we agree and merge this PR, I'll update wikis, create PRs to boilerplate generator, and other repos to match the same convention, to be ready to merge once we're happy with the GLA test-ride

<!--
Optional.
Enter a summary of all changes in this Pull Request, which will be added to the changelog if accepted.
Each line should start with `(Fix|Add|Tweak|Update) - `, for example:
> Fix - I took care of something that wasn't working.
> Add - I added something new that's pretty cool.
> Tweak - I made a small change.
> Update - I made big changes to something that wasn't broken.

Leave the "Changelog entry" header in place completely empty, without any summary if no changelog entry is needed.
If you remove the "Changelog entry" header, the title of Pull Request will be used as the changelog entry.
-->
### Changelog entry

> Dev - changed the changelog types list.

tomalec committed:
Add release yml

tomalec commented:
I also like the categories proposed by @eason9487 in https://github.com/woocommerce/grow/pull/16/files#diff-1965ba61cfb7646e192348cd54ed5c86b5327d054b39ea11e5c39cf2597950fe

Especially `docs` and `breaking`. However, I'm not sure how practically would be the granular distinction between `update`, `feature`, and `enhancement`? Plus, I'd not use the `[actions]` prefix, as such information is useful not only for actions.


I believe it would be beneficiary to unify, what we use as a branch prefix, GH `type: ` label, and what we state in the changelog.
Also, I wonder should we stick to the existing WP/WC-wide convention of prefixes, or take this opportunity, to introduce more types, and use categories instead of prefixes?


@eason9487 WDYT of making this tiny PR a place to decide on convention, and merge it sooner than later.
Then our PRs to all the tools around would be slightly easier to test and have one topic to discuss less.

tomalec committed:
Add emojis to categories and some whitespace,

to improve readability.

Inspiered by https://github.com/woocommerce/grow/commit/ec90f378c85df41a7a2344fc746eed2a4b2b114b

eason9487 commented:
> However, I'm not sure how practically would be the granular distinction between `update`, `feature`, and `enhancement`?

Considering the actions package in Grow repo is a tool, most of the categories are referenced from [Gutenberg's](https://github.com/WordPress/gutenberg/blob/v13.4.0/bin/plugin/commands/changelog.js#L160-L172). The mapping in my mind for the actions package is:

- feature - added a brand new feature.
- update - added a new function to an exist feature or updated an exist function.
- enhancement - made something better but didn't change its functionality such as performance enhancement; or added technical things like dependent upgrades.
- tweak - same as the proposal in this repo.

But I haven't thought about the comparison of these categories and the actions package probably doesn't need some of them. We can adjust them to have the same subset (fix, add, update, and tweak) as this repo, and add new categories when they are needed.

> Plus, I'd not use the `[actions]` prefix, as such information is useful not only for actions.

> I believe it would be beneficiary to unify, what we use as a branch prefix, GH `type: ` label, and what we state in the changelog.

I'm not pretty understand this. Do you mean the labels used by various packages in Grow repo should use the same label set in GLA?

Due to the monorepo, I thought the actions package needs the `[actions]` prefix to only pick out PRs that are related to actions package for generating the release notes. Otherwise, using the shared labels with other packages in Grow without prefix, the notes would include other PRs and it would be hard to filter unrelated titles out.

> Also, I wonder should we stick to the existing WP/WC-wide convention of prefixes, or take this opportunity, to introduce more types, and use categories instead of prefixes?

The prefix pattern is also referenced from Gutenberg. But I think it can be changed to other more suitable, as long as the other one is also identifiable as being dedicated to a specific package.

> WDYT of making this tiny PR a place to decide on convention, and merge it sooner than later.
> Then our PRs to all the tools around would be slightly easier to test and have one topic to discuss less.

Sounds good! 👍

tomalec commented:
> Due to the monorepo, I thought the actions package needs the [actions] prefix to only pick out PRs that are related to actions package for generating the release notes. Otherwise, using the shared labels with other packages in Grow without prefix, the notes would include other PRs and it would be hard to filter unrelated titles out.

Ah, thanks. Now I get it. Previously I thought `[actions] changelog: fix` denotes "This label is to be consumed by GH actions to set changelog category to 'fix'". But it's more like: "set changelog category to 'fix', for the actions package". And, I totally agree `[actions]` prefix is a good idea. Alternatively, we could have `package: actions` + `changelog: fix` not to multiply all the combinations, for all packages.

> > I believe it would be beneficiary to unify, what we use as a branch prefix, GH `type:` label, and what we state in the changelog.
> 
> I'm not pretty understand this. Do you mean the labels used by various packages in Grow repo should use the same label set in GLA?

Across all our repos, currently we do:
- Prefix branch names with `add/fix/update/remove`
- Prefix changelog entries with `Add/Fix/Update/Tweak`
- Attach GH label `type: bug/documentation/enhancement/task/technical debt *` (mostly to issues, but I tend to attach to PRs as well, especially those which does not have a corresponding issue)
- Your PR uses labels `[actions] type: breaking/feature/update/enhancement/fix/tweak/docs`
- Mine also use labels
IMO, All those things serve a single purpose: denote the type of change to be applied and to keep our life easier it would be better to use a single set of types/categories.

Yes, I do mean to use the same set of changelog-type-related labels for all our repos, and individual packages.

tomalec commented:
> > Also, I wonder should we stick to the existing WP/WC-wide convention of prefixes, or take this opportunity, to introduce more types, and use categories instead of prefixes?
> 
> The prefix pattern is also referenced from Gutenberg. But I think it can be changed to other more suitable, as long as the other one is also identifiable as being dedicated to a specific package.

Sorry, what I meant was not the package name prefix for the label. But the changelog entry prefixes, like in https://github.com/woocommerce/woocommerce/blob/trunk/changelog.txt#L38-L39
```
* Add - Add a WooCommerce > Subscriptions menu item for eligible stores to offer WC Payments Subscriptions functionality  [#32958](https://github.com/woocommerce/woocommerce/pull/32958)
* Dev - Add Preview store button to Home screen  [#32739](https://github.com/woocommerce/woocommerce/pull/32739)
* Tweak - For Vietnam, the second street address line should be displayed but not required.  [#32610](https://github.com/woocommerce/woocommerce/pull/32610)
* Update - Add ExPlat usage.  [#33023](https://github.com/woocommerce/woocommerce/pull/33023)
```
And here are actually two things to stick to or change:
1. Set of types/categories (scope of this PR)
2. :scroll:  Convention of 
	  ```
	  * {type} - {description}
	  * Fix - some bug
	  * Add - new feature
	  ```
	  vs.
	  ```
	  ### Fixed 🐛 
	   * Some bug
	  ### New Features 🎉
	   * New feature
	   ```
	   
Personally, I'd vote for the latter, as looking at https://github.com/woocommerce/woocommerce/blob/trunk/changelog.txt#L7-L141 the number of prefixes makes it harder to read and distinct the actual entries, but the (quite many) entries like "* **Add - Add** tracks to products …", "* **Dev - Add** dropins to WCTracker", "* **Dev - Fix** typescript type" makes it even more confusing.

eason9487 commented:
> Alternatively, we could have `package: actions` + `changelog: fix` not to multiply all the combinations, for all packages.

The matching PR rule of `labels` in .github/release.yml matches one of labels but does not match all. Probably we could not use combinations. Unless we want to set it up like this:

```yaml
changelog:
  categories:
    - title: New Features 🎉
      labels:
        - "package: actions"
        - "changelog: add"
      exclude:
        labels:
          - "package: storybook"
          - "package: jsdoc"
          - "changelog: fix"
          - "changelog: tweak"
          - # enumerate all other package and changelog labels
```

> * IMO, All those things serve a single purpose: denote the type of change to be applied and to keep our life easier it would be better to use a single set of types/categories.

Thanks for the explanation. Now I got the idea clearer. I agree with using the same set of prefix branch names to map the same set of prefix changelog entries or changelog titles. 👍 

But probably the package type repo would still need one more prefix like `break` for the major changes since the `remove` prefix might not so fit with some situations of bumping the major version.

eason9487 commented:
> 2\. 📜  Convention of
>    ```
>    * {type} - {description}
>    * Fix - some bug
>    * Add - new feature
>    ```
>    vs.
>    ```
>    ### Fixed 🐛 
>     * Some bug
>    ### New Features 🎉
>     * New feature
>    ```

The plugins published on WPORG extract the content of [the latest three versions in the changelog.txt](https://wordpress.org/plugins/google-listings-and-ads/#developers). And The plugins only published on WCCOM directly use [the changelog.txt as the Version history](https://woocommerce.com/products/automatewoo/). Both don't render the content as Markdown. Therefore, for the WC plugin type, I guess we need to stick with the current format for changelog.txt.

![image](https://user-images.githubusercontent.com/17420811/174532687-61ae064a-9c86-4a38-8b09-01bfc531ffa6.png)

Referring to the repos of [wp-components](https://github.com/WordPress/gutenberg/blob/%40wordpress/components%4019.13.0/packages/components/CHANGELOG.md) and [wc-components](https://github.com/woocommerce/woocommerce/blob/6.6.0/packages/js/components/CHANGELOG.md), they both have only CHANGELOG.md. So for the package type, using Markdown should be in line with the WP/WC convention.

> Personally, I'd vote for the latter, as looking at https://github.com/woocommerce/woocommerce/blob/trunk/changelog.txt#L7-L141 the number of prefixes makes it harder to read and distinct the actual entries, but the (quite many) entries like "* **Add - Add** tracks to products …", "* **Dev - Add** dropins to WCTracker", "* **Dev - Fix** typescript type" makes it even more confusing.

Assumed the audience of changelog.txt is plugin users. Perhaps it's a bit better to use the current format since the raw content should be easier for users to read than the Markdown. To reduce the confusion like **Add - Add** something, **Dev - Fix** something, it seems we don't have many choices from plugin repos. Maybe add more descriptions in the PR template to point out how the PR title will be transformed into a changelog entry.

I guess it should be ok to add CHANGELOG.md to plugin-type repos for the audience of developers, and it could have a more detailed changelog such as the internal changes. Then it would help track the history from the development perspective. Just to take care it won't be included in the plugin bundle in the private repos.

eason9487 commented:
A relevant idea about generating the format for changelog.txt. Because GitHub API only returns category titles from .github/release.yml with PR titles as entries, the `categories[*].title` in .github/release.yml could be:

```yaml
  categories:
    - title: "[Fix] Fixes 🛠"
      labels:
        - "changelog: fix"
        - "type: bug"

    - title: "[Add] New Features 🎉"
      labels:
        - "changelog: add"
        - "type: enhancement"
```

and GitHub API returns:

```markdown
## What's Changed
### [Add] New Features 🎉
* More promotion supported countries by @ianlin in https://github.com/woocommerce/google-listings-and-ads/pull/1536
### [Fix] Fixes 🛠
* CES Prompts Showing Several Times by @jorgemd24 in https://github.com/woocommerce/google-listings-and-ads/pull/1564
```

Then it would be easier to transform it to the following format without querying PRs:
```
* Add - More promotion supported countries.
* Fix - CES Prompts Showing Several Times.
```

💡 PS. In this example, I removed **Add** and **Fix -** from the actual titles of these PRs.

tomalec committed:
Exclude `github-actions` user from changelog,

Remove `[bot]` suffix from `dependabot`.

tomalec committed:
Add changelog entry prefix to the release notes category title

As suggested in https://github.com/woocommerce/google-listings-and-ads/pull/1565#issuecomment-1160102244

tomalec commented:
> The matching PR rule of labels in .github/release.yml matches one of labels but does not match all. Probably we could not use combinations.

Ah, fair point. Let's stick to `[package]` prefix (at least until we find a solution for that)

> But probably the package type repo would still need one more prefix like break for the major changes since the remove prefix might not so fit with some situations of bumping the major version.

Agree. I added `breaking` category here, but not `removed` (https://github.com/woocommerce/google-listings-and-ads/pull/1565/commits/0a2e9a95d8bd0e842463ec28d58b7acdabc178f3). I think if removal is a breaking change it's more important to highlight that's breaking, if it removes some legacy, unused code, it falls to update or tweak

tomalec commented:
> The plugins published on WPORG extract the content of [the latest three versions in the changelog.txt](https://wordpress.org/plugins/google-listings-and-ads/#developers). And The plugins only published on WCCOM directly use [the changelog.txt as the Version history](https://woocommerce.com/products/automatewoo/). Both don't render the content as Markdown. Therefore, for the WC plugin type, I guess we need to stick with the current format for changelog.txt.

Ok, let's stick to old format. 
I still think even in plain text some header would improve readability
```
= 6.5.1 2022-05-12 =

*** Fixed bugs ***

 * Make sure WooCommerce Admin is also being disabled on multi sites. [#32981](https://github.com/woocommerce/woocommerce/pull/32981)
 * Skip payment welcome screen for wc pay banner experiment [#33071](https://github.com/woocommerce/woocommerce/pull/33071)

…

*** Tweaks ***
 * Center experimental products view more button [#32870](https://github.com/woocommerce/woocommerce/pull/32870)
 * Refactor and improve tests payments task [#32662](https://github.com/woocommerce/woocommerce/pull/32662)
```

Especially, that they already parse some kind of headers/categories `** foo **` https://wordpress.org/plugins/woocommerce/#developers

But let's leave that for another iteration or issue.


>  Maybe add more descriptions in the PR template to point out how the PR title will be transformed into a changelog entry.
> 
> I guess it should be ok to add CHANGELOG.md to plugin-type repos for the audience of developers, and it could have a more detailed changelog such as the internal changes. 

:+1:

tomalec commented:
> A relevant idea about generating the format for changelog.txt. Because GitHub API only returns category titles from .github/release.yml with PR titles as entries, the categories[*].title in .github/release.yml could be:

Cool, done in https://github.com/woocommerce/google-listings-and-ads/pull/1565/commits/e173fb1d2bce3bb41354fdc07ced4a257278809a

tomalec committed:
Add Breaking and Documentation categories to changelog

tomalec commented:
Ok, so I propose here a following list of types/categories:

<table>
<tr>
	<th>Category name
	<th>Branch prefix
	<th>Changelog entry prefix
	<th>Semver level
	<th>GH Labels used
	<th>description
<tr>
	<td>[Break] Breaking Changes 🚨
	<td><code>breaking/</code>
	<td><code>Breaking – </code>
	<td>major
	<td><code>changelog: breaking</code>
	<td>There is a change breaking previous API or functionality, like removing some features, endpoints, etc.
<tr>
	<td>[Add] New Features 🎉
	<td><code>add/</code>
	<td><code>Add – </code>
	<td>minor
	<td><code>changelog: add</code>
	<td>A new feature, function or functionality was added.
<tr>
	<td>[Update] Updated ✨
	<td><code>update/</code>
	<td><code>Update – </code>
	<td>minor
	<td><code>changelog: update</code>
	<td>Big changes to something that wasn't broken.
<tr>
	<td>[Fix] Fixes 🛠
	<td><code>fix/</code>
	<td><code>Fix – </code>
	<td>patch
	<td><code>changelog: fix</code>
	<td>Took care of something that wasn't working.
<tr>
	<td>[Tweak] Tweaked 🔧
	<td><code>tweak/</code>
	<td><code>Tweak – </code>
	<td>patch
	<td><code>changelog: tweak</code>
	<td>Small change. A one that's technically affecting users, but isn't actually very important. Like a spelling fix, compatibility bump.
<tr>
	<td>[Dev] Developer-facing changes 🧑‍💻
	<td><code>dev/</code>
	<td><code>Dev – </code>
	<td>patch
	<td><code>changelog: dev</code>
	<td>Developer-facing only change, like refactoring, small tooling change.
<tr>
	<td>[Doc] Documentation 📚
	<td><code>doc/</code>
	<td><code>Doc – </code>
	<td>patch
	<td><code>changelog: doc</code>
	<td>Updated customer or developer facing documentation
</table>

- Category name - will be used only in developer-facing `changelog.md`
- Branch name - is just asuggested convention. Following it is not enforced anyhow, but would result in [automatic label assignment](https://github.com/woocommerce/grow/pull/14).
- Changelog entry prefix - is the thing that will eventually land in `changelog.txt`, used by WPORG, we will stick to it, and do not use any other prefixes
- Semver level - it's the level of severity used by automated tools to infer which version to bump
- GH Labels used - Github label to be set (manually or automatically) on PR to set given category/type.


------


Is there anything you would like to add, delete, change, shorten or reword here?

eason9487 commented:
Thanks for organizing the proposal. It looks good to me!

> * Changelog entry prefix - is the thing that will eventually land in `changelog.txt`, used by WPCOM, we will stick to it, and do not use any other prefixes

I think there is only one thing that needs an adjustment. The WPCOM in this description should be WPORG.

JPry commented:
> Is there anything you would like to add, delete, change, shorten or reword here?

I think we should also have a "Dev" prefix. You mentioned that this could be separate from the "Tweak" category, and I think it should be a separate category.

tomalec commented:
> I think we should also have a "Dev" prefix. You mentioned that this could be separate from the "Tweak" category, and I think it should be a separate category.

Personally, the shorter the list the better, so if we have two categories, they better be clearly distinct.

@JPry How would you describe them?
Also, when we move dev-related changes away from tweak, where would we draw the line between update and tweak?

<table>
<tr>
	<td>[Update] Updated ✨
	<td><code>update/</code>
	<td><code>Update – </code>
	<td>minor
	<td><code>changelog: update</code>
	<td>Big changes to something that wasn't broken.
<tr>
	<td>[Tweak] Tweaked 🔧
	<td><code>tweak/</code>
	<td><code>Tweak – </code>
	<td>patch
	<td><code>changelog: tweak</code>
	<td>Small change to something that wasn't broken.
<tr>
<tr>
	<td>[Dev] Dveloper-facing changes 🔧
	<td><code>dev/</code>
	<td><code>Dev – </code>
	<td>patch
	<td><code>changelog: dev</code>
	<td>Developer-facing change, like refactoring.
<tr>
</table>

JPry commented:
> @JPry How would you describe them?
> Also, when we move dev-related changes away from tweak, where would we draw the line between update and tweak?

In my mind, the distinction between a change labeled `dev` versus `tweak` is whether it will directly affect the end user experience in some way. One good example is updating Composer dependencies; this is something we need to do periodically, and it makes sense to indicate that it was done in the change log. But will the end user see some kind of change? No, so it would be a `dev` change. Put another way, `dev` should be a label that we use for notable things our team wants to track, or notable things that 3rd party developers might care about.

What makes a change a `tweak`? I think things like bumping version compatibility, or making small adjustments to text or styling, or other things that really don't make a big difference for the end user, but they are ultimately visible in some way.

tomalec commented:
Thanks, I updated https://github.com/woocommerce/google-listings-and-ads/pull/1565#issuecomment-1163223328, to 
<table>
<tr>
	<td>[Update] Updated ✨
	<td><code>update/</code>
	<td><code>Update – </code>
	<td>minor
	<td><code>changelog: update</code>
	<td>Big changes to something that wasn't broken.
<tr>
	<td>[Tweak] Tweaked 🔧
	<td><code>tweak/</code>
	<td><code>Tweak – </code>
	<td>patch
	<td><code>changelog: tweak</code>
	<td>Small change. A one that's technically affecting users, but isn't actually very important. Like a spelling fix, compatibility bump.
<tr>
	<td>[Dev] Developer-facing changes 🧑‍💻
	<td><code>dev/</code>
	<td><code>Dev – </code>
	<td>patch
	<td><code>changelog: dev</code>
	<td>Developer-facing only change, like refactoring, small tooling change.
	</table>

tomalec committed:
Add `dev` changelog type.

Addresses https://github.com/woocommerce/google-listings-and-ads/pull/1565#issuecomment-1164740447

tomalec committed:
Update PR template to reflect agreed chnagelog types

eason9487 reviewed (approved):
Thanks for the proposal. LGTM. I guess it's ready to proceed with the merge if all thoughts were settled down.

tomalec commented:
Updated the local wiki regarding branch name convention https://github.com/woocommerce/google-listings-and-ads/wiki/Working-With-Code/_compare/e9bb5d46a0f251097841ce1e6d29c76cbc59575a...5eba2ab58b2bd3c2a93f575cc14964dc74df32a9
And a note to Ventures P2 stating GLA experiment
